hrm-domain/hrm-service/src/contact/canPerformContactAction.ts

```diff
@@ -0,0 +1,166 @@
+
+const canPerformActionOnContact = (
+  action: (typeof actionsMaps.contact)[keyof typeof actionsMaps.contact],
+  additionalValidationForFinalized: (
+    contact: WithLegacyCategories<Contact>,
+    req: any,
+  ) => Promise<boolean> = () => Promise.resolve(true),
+) =>
+  asyncHandler(async (req, res, next) => {
+    if (!req.isAuthorized()) {
+      const { accountSid, user, can, body } = req;
+      const { contactId } = req.params;
+
+      try {
+        const contactObj = await getContactById(accountSid, contactId, {
+          can: req.can,
+          user,
+        });
+        if (!contactObj) {
+          // This is a dirty hack that relies on the catch block in the try/catch below to return a 404
+          throw new Error('contact not found');
+        }
+        if (contactObj.finalizedAt) {
+          if (
+            (await additionalValidationForFinalized(contactObj, req)) &&
+            can(user, action, contactObj)
+          ) {
+            req.authorize();
+          } else {
+            req.unauthorize();
+          }
+        } else {
+          // If there is no finalized date, then the contact is a draft and can only be edited by the worker who created it or the one who owns it.
+          // Offline contacts potentially need to be edited by a creator that won't own them.
+          // Transferred tasks need to be edited by an owner that didn't create them.
+          if (
+            contactObj.createdBy === user.workerSid ||
+            contactObj.twilioWorkerId === user.workerSid
+          ) {
+            req.authorize();
+          } else {
+            // It the contact record doesn't show this user as the contact owner, but Twilio shows that they are having the associated task transferred to them, permit the edit
+            // Long term it's wrong for HRM to be verifying this itself - we should probably initiate updates for the contact record from the backend transfer logic rather than Flex in future.
+            // Then HRM just needs to validate the request is coming from a valid backend service with permission to make the edit.
+            const twilioClient = await getClient({
+              accountSid,
+              authToken: process.env[`TWILIO_AUTH_TOKEN_${accountSid}`],
+            });
+            if (
+              await isTwilioTaskTransferTarget(
+                twilioClient,
+                body?.taskId,
+                contactObj.taskId,
+                user.workerSid,
+              )
+            ) {
+              req.authorize();
+            } else {
+              req.unauthorize();
+            }
+          }
+        }
+      } catch (err) {
+        if (
+          err instanceof Error &&
+          err.message.toLowerCase().includes('contact not found')
+        ) {
+          throw createError(404);
+        } else {
+          console.error('Failed to authorize contact editing', err);
+          throw createError(500);
+        }
```

Comment:
The Result type seems more set up for returning from route implementations rather than pipeline interceptors? The 'OK' result type needs a status code, but we don't want to return a status code in these functions, just call 'next()', so I wasn't sure it was the best fit?

Reply:
Not really, that was an introduction from Bo that we already agreed is not correct, but I didn't took the time to remove the status yet. We'll change to having an optional `name` in the error variants so we can have an "errors dictionary" and consume them when needed.